### ISSUE TITLE
Events in NodeController are not working

### ISSUE THREAD

**Issue by gmarek** [kubernetes/kubernetes#41318]
I noticed that we have no events from NodeController for some time now and I suspect this PR as this touches Event machinery. I see logs like:
```
Could not construct reference to: '&v1.ObjectReference{Kind:"Node", Namespace:"", Name:"gke-gke-large-deploy-default-pool-1-db034145-l952", UID:"a1c9fe9d-f1ce-11e6-ad7a-42010af00002", APIVersion:"", ResourceVersion:"", FieldPath:""}' due to: 'object does not implement the List interfaces'. Will not report event: 'Normal' 'RegisteredNode' 'Node gke-gke-large-deploy-default-pool-1-db034145-l952 event: Registered Node gke-gke-large-deploy-default-pool-1-db034145-l952 in NodeController'
```
I fixed the problem (using types from `pkg/api/v1` instead of `client-go/pkg/api/v1` when creating Event messages), but once I did this unit tests started to panic:
```
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x12c3983]

goroutine 14 [running]:
panic(0x22ad7c0, 0xc420012150)
	/usr/local/google/home/gmarek/.gvm/gos/go1.7.3/src/runtime/panic.go:500 +0x1ae
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x17d
panic(0x22ad7c0, 0xc420012150)
	/usr/local/google/home/gmarek/.gvm/gos/go1.7.3/src/runtime/panic.go:458 +0x271
k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*RESTClient).Verb(0x0, 0x248ac0c, 0x4, 0xc42011f501)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/client.go:222 +0x63
k8s.io/kubernetes/vendor/k8s.io/client-go/rest.(*RESTClient).Post(0x0, 0x1f8)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/rest/client.go:232 +0x4e
k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*events).CreateWithEventNamespace(0xc42019a600, 0xc420752200, 0xed033a0c6, 0x6, 0x43dd37)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go:53 +0xfe
k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*EventSinkImpl).Create(0xc4204f2460, 0xc420752200, 0xc42047a000, 0x30, 0x28)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go:155 +0x60
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.recordEvent(0x31a4160, 0xc4204f2460, 0xc420752200, 0x0, 0x0, 0x0, 0x405c00, 0xc42019a6a0, 0x2493bd7)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:180 +0xec
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.recordToSink(0x31a4160, 0xc4204f2460, 0xc4200fa000, 0xc42019a6a0, 0xc42019a640, 0x2540be400)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:138 +0x1c8
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartRecordingToSink.func1(0xc4200fa000)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:120 +0xb7
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartEventWatcher.func1(0x3199f60, 0xc4204dc7e0, 0xc4204dc7b0)
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:238 +0x127
created by k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartEventWatcher
	/usr/local/google/home/gmarek/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/tools/record/event.go:240 +0x91
```

cc @deads2k @wojtek-t @piosz

**Comment by deads2k** [comment]
There seem to be multiple node controllers now.  Which one is failing?

**Comment by deads2k** [comment]
Think I found it.  Helpers for helpers for helpers.  I'll open a pull shortly.

**Comment by gmarek** [comment]
I'm including the fix in #41311, so NC is covered. I'm not sure that the same thing wasn't happening in other components as well.

**Comment by deads2k** [comment]
Opened https://github.com/kubernetes/kubernetes/pull/41322 .

**Comment by deads2k** [comment]
> I'm including the fix in #41311, so NC is covered. I'm not sure that the same thing wasn't happening in other components as well.

Ah, I didn't see the fix I expected there, so #41322 may still be needed.  The other controllers pass their actual object through, not an `ObjectReference` that is getting matched by type assertion, so they're ok.  Seem like only kubelet related things do that, so I've picked up a couple stragglers after the transition.

Just to briefly explain why we have this pain.  All controllers, the scheduler, and the kubelet are supposed to transition to client-go.  We're working to make that transition easy, though we may not actually hit switch ourselves.  We needed to eliminate the incompatible copies for the  EventRecorder and collapse the code paths down, so we made client-go authoritative since its the final destination.
